### PR TITLE
Adds output address setting to DLA.py

### PR DESCRIPTION
**Pull request by vilukissa68**
Enables "hidden register" 5c which is used for setting the output address for the DLA. 
Includes accumulated minor fixes and changes.

### CODE REVIEW

File: vp/devel/python_peripherals/DLA.py [hunk 1]
```diff
@@ -248,15 +255,22 @@ def clip(value, clip, no_overflow=False):
...
-        if no_overflow:
-            value
-        return (value, 0)
+        if value > mask:
+                return mask-1 if no_overflow else (mask-1, value - mask)
+        elif value < -mask:
+                return -mask if no_overflow else (-mask, value + mask)
+
+    return value if no_overflow else (value, 0)
+
```
Comment (hegza) [inline vp/devel/python_peripherals/DLA.py]: The control flow of this function gets a bit hard to follow. Consider splitting to `clip_signed` and `clip_unsigned`.

Comment (hegza) [reply vp/devel/python_peripherals/DLA.py]: Good rule of thumb: if you're going to your third level of indentation, reconsider what you're doing.

File: vp/devel/python_peripherals/DLA.py [hunk 2]
```diff
@@ -248,15 +255,22 @@ def clip(value, clip, no_overflow=False):
...
+            return mask if no_overflow else (mask, value - mask)
+        elif value < 0:
+            return 0 if no_overflow else (0, value)
+
     else:
-        if no_overflow:
-            value
-        return (value, 0)
+        if value > mask:
+                return mask-1 if no_overflow else (mask-1, value - mask)
```
Comment (hegza) [inline vp/devel/python_peripherals/DLA.py]: Yo, inconsistent indentation.

File: vp/devel/python_peripherals/DLA.py [hunk 3]
```diff
@@ -300,30 +305,26 @@ def write(self, data):
...
+        data -- [Int] data to write to bank
 
-        Return:
-        free -- Int Free space in bank
+        Returns:
+        unwritten -- [Int] data that didn't fit to bank
         """
-        return self.size - self.idx
+        assert(offset < self.size)
+        self.mem[offset] = data & 0xFF # Enforce 8bit width
```
Comment (hegza) [inline vp/devel/python_peripherals/DLA.py]: Nice doc, I can read this.

However, I think it's not entirely accurate in this case. Looks stale :)

Comment (hegza) [reply vp/devel/python_peripherals/DLA.py]: To be clear, this function does not return anything, unlike what is claimed.

File: vp/devel/python_peripherals/DLA.py [hunk 4]
```diff
@@ -300,30 +305,26 @@ def write(self, data):
...
+
 
+    def read(self, offset):
```
Comment (hegza) [inline vp/devel/python_peripherals/DLA.py:331]: By the way should these have some more expressive name such as `mmio_read` or `dla_read`? I think the `read` is a bit too generic for how these are used.

File: vp/devel/python_peripherals/DLA.py [hunk 5]
```diff
@@ -435,68 +451,101 @@ def handle_bank_write(self, request):
...
-                    slice = data[i:]
-                    i = len(data)
-                else:
-                    slice = data[i:i+bank.available_memory()]
-                    i = i + bank.available_memory()
-                bank.write(slice)
-        if i != len(data):
-            return data[i:]
-        return []
+                print("Output written outside memory region")
```
Comment (hegza) [inline vp/devel/python_peripherals/DLA.py]: Is this an error, warn or trace? Consider tagging it as appropriate, e.g. `print("ERROR: output written outside memory region)`.

File: vp/devel/python_peripherals/DLA.py [hunk 6]
```diff
@@ -604,18 +658,34 @@ def pp_clip(self, values):
         """Clip pp values if register is set"""
         clip_amount = self.get_register(PP_CTRL, PP_CLIP_OFFSET, 5)
         if clip_amount > 0:
-            return execute_for_all_elements(clip, values, clip_amount, True)
+            return execute_for_all_elements(clip, values, clip_amount, True, True)
```
Comment (hegza) [inline vp/devel/python_peripherals/DLA.py]: Pretty cool, this higher order function ':D

Comment (hegza) [reply vp/devel/python_peripherals/DLA.py]: I wonder if there's a map operation for Python 3.4. It might be more readable.

**Review by hegza — CHANGES_REQUESTED** [review]
Minor changes.

I'm assuming behavior is tested correct. We might want to review this more holistically at some point but I'm thinking it looks pretty good, considering available tooling.

File: vp/devel/python_peripherals/DLA.py (hunk 4 above)
Comment (vilukissa68) [reply vp/devel/python_peripherals/DLA.py:331]: Atleast that particular function is a method in the MemoryBank class. So the usage is always `bankobj.read(addr)`. Personally I think that this makes more sense than `bankobj.bank_read(addr)`, since there is only one kind of memory one can read from the bank.

File: vp/devel/python_peripherals/DLA.py (hunk 6 above)
Comment (vilukissa68) [reply vp/devel/python_peripherals/DLA.py]: It might have been that 3.4 Map() couldn't handle nested lists easily. I did try using it, but went with this fun programming exercise instead.

File: vp/devel/python_peripherals/DLA.py (hunk 1 above)
Comment (vilukissa68) [reply vp/devel/python_peripherals/DLA.py]: That's a good rule! Function splitted.

File: vp/devel/python_peripherals/DLA.py (hunk 2 above)
Comment (vilukissa68) [reply vp/devel/python_peripherals/DLA.py]: Don't know what happened here, probably some emacs shenanigans. Fixed indentation.

File: vp/devel/python_peripherals/DLA.py (hunk 3 above)
Comment (vilukissa68) [reply vp/devel/python_peripherals/DLA.py]: Documentation fixed

File: vp/devel/python_peripherals/DLA.py (hunk 5 above)
Comment (vilukissa68) [reply vp/devel/python_peripherals/DLA.py]: Tagged as a warning. This is indented functionality on the DLA but is not possible in VP due to it not having free access to the SDRAM.

File: vp/devel/python_peripherals/DLA.py (hunk 4 above)
Comment (hegza) [reply vp/devel/python_peripherals/DLA.py:331]: True, sounds good to me.

**Review by hegza — APPROVED** [review]
I think it's good enough. Let's do a more holistic review of the VP later.
